ansible/roles/vm_set/files/mux_simulator.md

```diff
@@ -0,0 +1,173 @@
+# Mux Simulator
+
+In dualtor testbed, mux Y cable is simulated by OVS in test server. An OVS bridge is created for each of the mux. The PTF interface and test server VLAN interfaces are attached to the bridge. The VLAN interfaces are connected to DUT ports through fanout switches.
+
+```
+                          +--------------+
+                          |              +----- upper_if
+          PTF (host_if) --+  OVS bridge  |
+                          |              +----- lower_if
+                          +--------------+
+```
+
+Open flow rules are configured for the OVS bridge to simulate upstream broadcasting and downstream dropping traffic from standby side.
+
+To further simulate mux Y cable active/standby querying and setting, a process need to be started in the test server. The process needs to expose APIs for querying and setting active/standby status. On DUT side, a plugin can be injected to intercept the calls to mux Y cable. Instead of calling the actual mux driver functions, the APIs of mux simulator are called. Then the process checks and updates open flow configurations of the OVS bridge accordingly
+
+The mux_simulator.py script is for such purpose. It is a [Flask](https://flask.palletsprojects.com/en/1.1.x/) based program exposing HTTP API. While running `testbed-cli.sh add-topo`, it is deployed to test server and started as a systemd service. By default, it listens on port `8080'.
+
+On SONiC DUT side, script [y_cable_simulator_client.py](https://github.com/Azure/sonic-mgmt/blob/master/ansible/dualtor/y_cable_simulator_client.j2) will be injected to DUT during `testbed-cli.sh deploy-mg`. The [y_cable driver](https://github.com/Azure/sonic-platform-common/blob/master/sonic_y_cable/y_cable.py) has been enhanced to call the simulator client if it is available.
+
+## Configuration
+The default TCP port that the mux simulator will be listening on is configurable by variable `mux_simulator_port` in [https://github.com/Azure/sonic-mgmt/blob/master/ansible/group_vars/all/variables](https://github.com/Azure/sonic-mgmt/blob/master/ansible/group_vars/all/variables).
+
+The mux simulator would be deployed to test server as a systemd server `mux-simulator` during `testbed-cli.sh add-topo`.
+```
+azure@test-server:~/veos-vm$ sudo systemctl status mux-simulator
+鈼 mux-simulator.service - mux simulator
+   Loaded: loaded (/etc/systemd/system/mux-simulator.service; static; vendor preset: enabled)
+   Active: active (running) since Fri 2020-12-25 07:56:21 UTC; 12s ago
+ Main PID: 928 (python)
+    Tasks: 1 (limit: 11059)
+   CGroup: /system.slice/mux-simulator.service
```

Comment:
There are some garbled character. Could you have a recheck?

Reply:
Sure, I'll update.